purchase_blanket_order/models/purchase_order.py

```diff
@@ -0,0 +1,168 @@
+                    )
+                    % order.name
+                )
+        return res
+
+    @api.constrains("partner_id")
+    def check_partner_id(self):
+        for line in self.order_line:
+            if line.blanket_order_line:
+                if line.blanket_order_line.partner_id != self.partner_id:
+                    raise ValidationError(
+                        _(
+                            "The vendor must be equal to the blanket order"
+                            " lines vendor"
+                        )
+                    )
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    blanket_order_line = fields.Many2one(
+        "purchase.blanket.order.line", "Blanket Order Line", copy=False
+    )
+
+    def _get_assigned_bo_line(self, bo_lines):
+        # We get the blanket order line with enough quantity and closest
+        # scheduled date
+        assigned_bo_line = False
+        date_planned = fields.Date.from_string(self.date_planned) or date.today()
+        date_delta = timedelta(days=365)
+        for line in bo_lines.filtered(lambda l: l.date_schedule):
+            date_schedule = fields.Date.from_string(line.date_schedule)
+            if date_schedule and abs(date_schedule - date_planned) < date_delta:
+                assigned_bo_line = line
+                date_delta = abs(date_schedule - date_planned)
+        if assigned_bo_line:
+            return assigned_bo_line
+        non_date_bo_lines = bo_lines.filtered(lambda l: not l.date_schedule)
+        if non_date_bo_lines:
+            return non_date_bo_lines[0]
+
+    def _get_eligible_bo_lines_domain(self, base_qty):
+        filters = [
+            ("product_id", "=", self.product_id.id),
+            ("remaining_qty", ">=", base_qty),
+            ("currency_id", "=", self.order_id.currency_id.id),
+            ("order_id.state", "=", "open"),
+        ]
+        if self.order_id.partner_id:
+            filters.append(("partner_id", "=", self.order_id.partner_id.id))
+        return filters
+
+    def _get_eligible_bo_lines(self):
+        base_qty = self.product_uom._compute_quantity(
+            self.product_qty, self.product_id.uom_id
+        )
+        filters = self._get_eligible_bo_lines_domain(base_qty)
+        return self.env["purchase.blanket.order.line"].search(filters)
+
+    def get_assigned_bo_line(self):
+        self.ensure_one()
+        eligible_bo_lines = self._get_eligible_bo_lines()
+        if eligible_bo_lines:
+            if (
+                not self.blanket_order_line
+                or self.blanket_order_line not in eligible_bo_lines
+            ):
+                self.blanket_order_line = self._get_assigned_bo_line(eligible_bo_lines)
+        else:
+            self.blanket_order_line = False
+        self.onchange_blanket_order_line()
+        return {"domain": {"blanket_order_line": [("id", "in", eligible_bo_lines.ids)]}}
+
+    @api.onchange("product_id", "partner_id")
```

Comment:
```suggestion
    @api.onchange( "partner_id")
```
AFAIR product_id assigned in super method so there is no need to add it here